ext/standard/hrtime.c

```diff
@@ -0,0 +1,206 @@
+/*
+   +----------------------------------------------------------------------+
+   | PHP Version 7                                                        |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 1997-2017 The PHP Group                                |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Niklas Keller <kelunik@php.net>                              |
+   +----------------------------------------------------------------------+
+ */
+
+/* $Id$ */
+
+#include "php.h"
+#include "zend_exceptions.h"
+#include "hrtime.h"
+
+/* {{{ */
+/* This file reuses code parts from the cross-platform timer library
+	Public Domain - 2011 Mattias Jansson / Rampant Pixels */
+
+#if PHP_HRTIME_PLATFORM_POSIX
+
+# include <unistd.h>
+# include <time.h>
+# include <string.h>
+
+#elif PHP_HRTIME_PLATFORM_WINDOWS
+
+# define WIN32_LEAN_AND_MEAN
+# include <windows.h>
+
+static uint64_t _timer_freq = 0;
+static double _timer_int = .0;
+
+#elif PHP_HRTIME_PLATFORM_APPLE
+
+# include <mach/mach_time.h>
+# include <string.h>
+static mach_timebase_info_data_t _timerlib_info;
+
+#elif PHP_HRTIME_PLATFORM_HPUX
+
+# include <sys/time.h>
+
+#elif PHP_HRTIME_PLATFORM_AIX
+
+# include <sys/time.h>
+# include <sys/systemcfg.h>
+
+#endif
+
+#define NANO_IN_SEC 1000000000
+/* }}} */
+
+static int _timer_init()
+{/*{{{*/
+#if PHP_HRTIME_PLATFORM_WINDOWS
+
+	uint64_t unused;
+	if (!QueryPerformanceFrequency((LARGE_INTEGER*) &_timer_freq) ||
+		!QueryPerformanceCounter((LARGE_INTEGER*) &unused)) {
```

Comment:
It's merely to check whether the call is successful, as QueryPerformanceCounter will be used in further. At this point only what matters is to determine the counter frequency. Same applies to clock_gettime() and other places. It is a "probe" call to indicate the APIs work as expected at the startup time already.

Reply:
I'm not sure what the point to check it's successful here? First of all, MS docs guarantee that on any supported Windows system it always succeeds: https://msdn.microsoft.com/en-us/library/windows/desktop/ms644904(v=vs.85).aspx but if that proves not to be true, it could as well fail next time we call it and not this time?

Reply:
Oh, what they say is even "On systems that run Windows XP or later, the function will always succeed and will thus never return zero.". Odd, no way one can know it actually failed, and same about QueryPerformanceFrequency. Reworking this part then.